File: 1_Мod/dependencies.py
import sys


class Dependency:
    def __init__(self, name, parent=None, is_project=False):
        self.name = name
        self.parent = set()
        self.is_project = is_project
        if parent:
            self.parent = {parent}

    def add_parent(self, parent):
        self.parent.add(parent)

    def get_path(self, current, path):  # c 3.7 dict хранит порядок добавления

        if current.is_project:
            print(" ".join(list(path.keys())[::-1]))
            return

        if not current.parent or path.get(current.name):
            return

        path[current.name] = 1

        for x in current.parent:
            self.get_path(x, path)
        path.pop(current.name)


class Graph:
    def __init__(self, head, vulnerable):
        self.nodes = dict()
        self.head = head
        self.vulnerable = set(vulnerable)

    def get_child(self, name, parent):
        child = self.nodes.get(name)
        if child:
            child.add_parent(parent)
            return
        self.nodes[name] = Dependency(name, parent=parent)

    def get_parent(self, name):
        parent = self.nodes.get(name)
        if parent:
            return parent
        parent = Dependency(name)
        self.nodes[name] = parent
        return parent

    def print_paths(self):
        for name in self.vulnerable:
            if not self.nodes.get(name):
                continue
            self.nodes[name].get_path(self.nodes[name], dict())


def main():
    #test = open("C_test/test.txt", "r")
    vulnerable = input().split()  # input().split()
    line = input().split()  # input().split()
    graph = Graph(Dependency("project", is_project=True), vulnerable)

    for i in line:
        graph.get_child(i, graph.head)

    for line in sys.stdin:
        line = line.replace('\n', '').split()
        if not line:
            continue
        parent = graph.get_parent(line[0])

        for name in line[1:]:
            graph.get_child(name, parent)

    graph.print_paths()
    # test.close()


if __name__ == '__main__':
    main()

File: 1_Мod/dependencies.rs
use std::collections::{HashSet, HashMap};
use std::io;

struct Dependency{
    name: String,
    parents: HashSet<String>,
    is_project_dep: bool
}

impl Dependency {
    fn new(name: String, is_project_dep: bool) -> Self {
        Self {
            name,
            parents: HashSet::new(),
            is_project_dep
        }
    }

    fn add_parent(&mut self, parent: &String){
        self.parents.insert(parent.clone());
    }

}


struct Graph{
    nodes: HashMap<String, Dependency>
}

impl Graph{
    fn new() -> Self {
        Self{
            nodes: HashMap::new(),
        }
    }

    fn get_path(&self, node: &Dependency, path: &mut Vec<String>){


            if  node.is_project_dep {
                print!("{}", node.name);
                for dep in path.iter().rev(){
                    print!(" ");
                    print!("{}", dep);
                }
                println!();
            }

            if node.parents.is_empty() || path.contains(&node.name){
                return;
            }

            path.push(node.name.clone());
            let mut parent_node;
            for parent_name in node.parents.iter(){
                parent_node = self.nodes.get(parent_name);
                if parent_node.is_some(){
                    self.get_path(&parent_node.unwrap(), path);
                }
            }
            path.pop();
    }

    fn get_parent(&mut self, name: &String){
        let _parent = self.nodes.get(name);
        if !_parent.is_some(){
            self.nodes.insert(name.clone(), Dependency::new(name.clone(), false));
        }
    }

    fn get_child(&mut self, name: &String, parent_name: &String, is_project: bool){
        let _child = self.nodes.get_mut(name);
        match _child {
            Some(child) => {
                child.add_parent(parent_name);
            },
            _ => {
                let mut new = Dependency::new(name.clone(), is_project);
                new.add_parent(parent_name);
                self.nodes.insert(name.clone(), new);
            }
        }
    }
}

fn main(){
    let mut vulnerable = String::new();
    let mut project_dep = String::new();
    io::stdin().read_line(&mut vulnerable).unwrap();
    io::stdin().read_line(&mut project_dep).unwrap();
    let vuln_vec = vulnerable.split(" ").map(|x| x.to_owned()).collect::<Vec<String>>();
    let project_dep_vec = project_dep.split(" ").map(|x| x.to_owned()).collect::<Vec<String>>();

    let mut graph = Graph::new();

    for dep in project_dep_vec.iter(){
        graph.get_child(dep, &"project".to_string(), true);
    }

    let mut dep_vec: Vec<String>;
    for line in io::stdin().lines() {

        dep_vec = line.unwrap().split(" ").map(|x| x.to_owned()).collect::<Vec<String>>();
        graph.get_parent(&dep_vec[0]);
        for i in 1..dep_vec.len(){
            graph.get_child(&dep_vec[i], &dep_vec[0], false);
        }
    }
    let mut vuln;
    for vuln_name in vuln_vec.iter(){
        vuln = graph.nodes.get(vuln_name);
        if vuln.is_some(){
            graph.get_path(vuln.unwrap(),&mut Vec::new());
        }
    }

}